docs/spec/constructors.rst

```diff
@@ -0,0 +1,375 @@
+Constructors
+============
+
+Calls to constructors require special handling within type checkers.
+
+Constructor Calls
+-----------------
+
+At runtime, a call to a class' constructor typically results in the invocation of
+three methods in the following order:
+1. The ``__call__`` method of the metaclass (which is typically supplied by the
+  ``type`` class but can be overridden by a custom metaclass)
+2. The ``__new__`` static method of the class
+3. The ``__init__`` instance method of the class
+
+Type checkers should mirror this runtime behavior when analyzing a constructor
+call.
+
+Metaclass ``__call__`` Method
+=============================
+
+When evaluating a constructor call, a type checker should first check if the
+class has a custom metaclass (a subclass of ``type``) that defines a ``__call__``
+method. If so, it should evaluate the call of this method using the supplied
+arguments. If the metaclass is ``type``, this step can be skipped.
+
+If the evaluated return type of the ``__call__`` method is something other than
+``Any`` or an instance of the class being constructed, a type checker should
+assume that the metaclass ``__call__`` method is overriding ``type.__call__``
+in some special manner, and it should not attempt to evaluate the ``__new__``
+or ``__init__`` methods on the class. For example, some metaclass ``__call__``
+methods are annotated to return ``NoReturn`` to indicate that constructor
+calls are not supported for that class.
+
+  ::
+
+    class Meta(type):
+        def __call__(cls, *args, **kwargs) -> NoReturn:
+            raise TypeError("Cannot instantiate class")
+
+    class MyClass(metaclass=Meta):
+        def __new__(cls, *args, **kwargs) -> Self:
+            return super().__new__(cls, *args, **kwargs)
+
+    assert_type(MyClass(), Never)
+
+
+``__new__`` Method
+==================
+
+After the metaclass ``__call__`` method has been evaluated, a type checker
```

Comment:
```suggestion
After the metaclass ``__call__`` method has been evaluated (and if it isn't overridden in some special manner), a type checker
```

As written in the `__call__` section:
> a type checker should assume that the metaclass `__call__` method is overriding `type.__call__` in some special manner, and it should not attempt to evaluate the `__new__` or `__init__`

I think this makes things a bit less confusing